retry/api.py

```diff
@@ -1,84 +1,215 @@
+import asyncio
 import logging
 import random
-import time
-
+import traceback
+import threading
 from functools import partial
 
-from .compat import decorator
+import inspect
 
+from .compat import decorator
 
 logging_logger = logging.getLogger(__name__)
 
+class ConditionRelease(object):
+    def __init__(self, c):
+        self.c = c
+    def __enter__(self):
+        self.c.acquire()
+        return self
+    def wait(self, n):
+        self.c.wait(n)
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        self.c.release()
 
-def __retry_internal(f, exceptions=Exception, tries=-1, delay=0, max_delay=None, backoff=1, jitter=0,
-                     logger=logging_logger):
-    """
-    Executes a function and retries it if it failed.
-
-    :param f: the function to execute.
-    :param exceptions: an exception or a tuple of exceptions to catch. default: Exception.
-    :param tries: the maximum number of attempts. default: -1 (infinite).
-    :param delay: initial delay between attempts. default: 0.
-    :param max_delay: the maximum value of delay. default: None (no limit).
-    :param backoff: multiplier applied to delay between attempts. default: 1 (no backoff).
-    :param jitter: extra seconds added to delay between attempts. default: 0.
-                   fixed if a number, random if a range tuple (min, max)
-    :param logger: logger.warning(fmt, error, delay) will be called on failed attempts.
-                   default: retry.logging_logger. if None, logging is disabled.
-    :returns: the result of the f function.
-    """
+def __retry_internal(
+    f,
+    exceptions=Exception,
+    tries=-1,
+    delay=0,
+    max_delay=None,
+    backoff=1,
+    jitter=0,
+    show_traceback=False,
+    logger=logging_logger,
+    fail_callback=None,
+    condition=None,
+):
     _tries, _delay = tries, delay
+
     while _tries:
         try:
             return f()
+
+        except exceptions as e:
+            _tries -= 1
+
+            if logger:
+                _log_attempt(tries, show_traceback, logger, _tries, _delay, e)
+
+            if not _tries:
+                raise
+
+            if fail_callback is not None:
+                fail_callback(e)
+
+            with ConditionRelease(condition) as _condition:
+                print('_condtion:', _condition)
```

Comment:
Does it really need to print to stdout or did you want to add some logging?

Reply:
removing.